staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+// SchemaResolver finds the OpenAPI schema for the given GroupVersionKind.
+// This interface uses the type defined by k8s.io/kube-openapi
+type SchemaResolver interface {
+	// ResolveSchema takes a GroupVersionKind (GVK) and returns the OpenAPI schema
+	// identified by the GVK.
+	// The function returns a non-nil error if the schema cannot be found or fail
+	// to resolve. The returned error wraps ErrSchemaNotFound if the resolution is
+	// attempted but the corresponding schema cannot be found.
+	ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error)
+}
+
+// ErrSchemaNotFound is wrapped and returned if the schema cannot be located
+// by the resolver.
+var ErrSchemaNotFound = fmt.Errorf("schema not found")
+
+// ClientDiscoveryResolver uses client-go discovery to resolve schemas at run time.
+type ClientDiscoveryResolver struct {
+	Discovery discovery.DiscoveryInterface
+}
+
+var _ SchemaResolver = (*ClientDiscoveryResolver)(nil)
+
+func (r *ClientDiscoveryResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	p, err := r.Discovery.OpenAPIV3().Paths()
+	if err != nil {
+		return nil, err
+	}
+	resourcePath := resourcePathFromGV(gvk.GroupVersion())
+	c, ok := p[resourcePath]
+	if !ok {
+		return nil, fmt.Errorf("cannot resolve group version %q: %w", gvk.GroupVersion(), ErrSchemaNotFound)
+	}
+	b, err := c.Schema(runtime.ContentTypeJSON)
```

Comment:
how often is this called, and in what code path? are we expecting an implementation of the discovery interface that handles caching, etc?

Reply:
I expect resolved schemas to be cached. For now the resolution would happen for every type check, and, in worst case, for every evaluation. Performance tuning would happen when the feature graduates, however I do not expect this library to change as much. Instead, the results should be cached by the caller.

Yes we do expect the caller to do the caching and use the cached Discovery interface. The main reason is that the admission controller has knowledge of CRDs being changed (with an informer) and should be easier to invalidate.

Reply:
An API server process that needs schema information for a given GVK in order to do type-aware compilation for admission should not need to make a network call to resolve it, or download an openapi schema, or unmarshal it. 

The API server running admission already has the schema information in-process, it should use that rather than pulling from discovery.

Modeling this as a SchemaResolver interface is good, but I would have expected the first two implementations to be one that pulled from compiled-in openapi info (used by kube-apiserver for all its types, kube-aggregator for APIService, apiextensions-apiserver for CustomResourceDefinition and Scale types) and one that converted from CRD-defined schemas (used by apiextensions-apiserver for custom resources with schemas).

We don't need an API server process to evaluate a schema fetched from the network for a type served by another API server, so let's not put that complexity and performance hit in the admission path.

Reply:
> The API server running admission already has the schema information in-process, it should use that rather than pulling from discovery.

> We don't need an API server process to evaluate a schema fetched from the network for a type served by another API server, so let's not put that complexity and performance hit in the admission path.

This is resolved by the compiled-in definition resolver; the discovery resolver is however kept for client-side where the compiled-in schemas do not exist.

`one that converted from CRD-defined schemas (used by apiextensions-apiserver for custom resources with schemas).`

CRD to Structural is already there, if any change is needed I would consider it out of scope of this PR.